packages/contractkit/src/wallets/signers/azure-key-vault-client.ts

```diff
@@ -0,0 +1,181 @@
+      keyNames.push(keyProperties.name)
+    }
+    return keyNames
+  }
+
+  public async getPublicKey(keyName: string): Promise<BN> {
+    const signingKey = await this.getKey(keyName)
+
+    const pubKeyPrefix = Buffer.from(new Uint8Array([this.publicKeyPrefix]))
+    const rawPublicKey = Buffer.concat([
+      pubKeyPrefix,
+      Buffer.from(signingKey.key!.x!),
+      Buffer.from(signingKey.key!.y!),
+    ])
+    const publicKey = new BN(rawPublicKey)
+    return publicKey
+  }
+
+  public async getKeyId(keyName: string): Promise<string> {
+    if (!(await this.hasKey(keyName))) {
+      throw new Error(`Unable to locate key: ${keyName}`)
+    }
+    return (await this.getKey(keyName)).id!
+  }
+
+  public async signMessage(message: Buffer, keyName: string): Promise<Signature> {
+    if (!(await this.hasKey(keyName))) {
+      throw new Error(`Unable to locate key: ${keyName}`)
+    }
+    const keyId = await this.getKeyId(keyName)
+    const cryptographyClient = new CryptographyClient(keyId, this.credential)
+    // @ts-ignore-next-line (ECDSA256 is not included in the client enum but is valid)
+    const signResult = await cryptographyClient.sign(this.SIGNING_ALGORITHM, message)
+    // The output of this will be a 64 byte array.
+    // The first 32 are the value for R and the rest is S
+    if (
+      typeof signResult == 'undefined' ||
+      typeof signResult.result == 'undefined' ||
+      signResult.result.length != 64
+    ) {
+      throw new Error(`Invalid signature returned from Azure: ${signResult}`)
+    }
+    const rawSignature = signResult.result
+
+    // Canonicalize signature
+    const secp256k1Curve = new EC('secp256k1')
+    const R = new BN(Buffer.from(rawSignature.slice(0, 32)))
+    let S = new BN(Buffer.from(rawSignature.slice(32, 64)))
+
+    // The Azure Signature MAY not be canonical, which is illegal in Ethereum
+    // thus it must be transposed to the lower intersection.
+    // https://github.com/bitcoin/bips/blob/master/bip-0062.mediawiki#Low_S_values_in_signatures
+    if (!AzureKeyVaultClient.isCanonical(S, secp256k1Curve.curve.n)) {
+      console.log('not canon')
+      S = secp256k1Curve.curve.n.sub(S)
+    }
+
+    const canonicalizedSignature = Buffer.concat([R.toBuffer('be', 32), S.toBuffer('be', 32)])
+    const publicKey = await this.getPublicKey(keyName)
+
+    // Azure doesn't provide the recovery key in the signature
+    const recoveryParam = AzureKeyVaultClient.recoverKeyIndex(
+      canonicalizedSignature,
+      publicKey,
+      message
+    )
+    return new Signature(recoveryParam, R, S)
+  }
+
+  public async hasKey(keyName: string): Promise<boolean> {
+    try {
+      await this.keyClient.getKey(keyName!)
+    } catch (e) {
+      if (e.message.includes('this is not a valid private key')) {
+        return false
+      }
+      throw e
+    }
+    return true
+  }
+
+  /**
+   * Returns true if the signature is in the "bottom" of the curve
```

Comment:
What does `"bottom" of  the curve` mean, and how does that map to the concept of being "Canonical"?

Reply:
It has to do with the ECDSA curve and the standard that Eth (and BTC) uses. I added this link in the comment above: https://github.com/bitcoin/bips/blob/master/bip-0062.mediawiki#Low_S_values_in_signatures

If you're curious to dive deep on ECDSA signatures, I found this primer to be helpful:
https://blog.cloudflare.com/a-relatively-easy-to-understand-primer-on-elliptic-curve-cryptography/